### PR TITLE
feat(node): add relay/hole-punching functionality using multiple transport options (tcp/quic)

### PR DESCRIPTION
**Pull request by aed900**
Add relay/hole-punching functionality using multiple transport options (tcp/quic).

### CODE REVIEW

File: sn_networking/src/cmd.rs [hunk 1]
```diff
@@ -361,16 +361,21 @@ impl SwarmDriver {
...
-                .condition(PeerCondition::NotDialing)
-                .addresses(vec![addr])
-                .build(),
-            None => DialOpts::unknown_peer_id().address(addr).build(),
-        };
-
-        self.swarm.dial(opts)
+        if addr.to_string().contains("p2p-circuit") {
+            info!("Dialing manually relay address: {}", addr.clone());
+            self.swarm.dial(addr.clone())
```
Comment (joshuef) [inline sn_networking/src/cmd.rs]: do we want to avoid all the conditions below? Why?

File: sn_networking/src/event.rs [hunk 2]
```diff
@@ -234,6 +302,16 @@ impl SwarmDriver {
                 }
             },
             SwarmEvent::NewListenAddr { address, .. } => {
+                if address.iter().any(|p| p == libp2p::multiaddr::Protocol::P2pCircuit) {
+                    info!("Relayed listen address added: {}", address.clone());
```
Comment (joshuef) [inline sn_networking/src/event.rs:331]: lets make this a Marker

**Comment by joshuef** [comment]
Why do we have relay and hole punching together?

Lets get these split out into separate PRs so we can verify them independently. 

Same for TCP/quic. I want to know we're using a given protocol for messaging or not. Lets not mix all these changes together.

File: sn_networking/src/event.rs (hunk 2 above)
Comment (aed900) [reply sn_networking/src/event.rs:331]: Which is more helpful here?
RelayedListenAddressAdded
RelayReservationAccepted

And should I add another for RelayedConnectionEstablished in the ConnectionEstablished event?

File: sn_networking/src/cmd.rs (hunk 1 above)
Comment (aed900) [reply sn_networking/src/cmd.rs]: This was for debugging some dial issues. Removed now.

**Comment by aed900** [comment]
Split into separate PRs:
TCP/QUIC feat flag: https://github.com/maidsafe/safe_network/pull/630
Relay: https://github.com/maidsafe/safe_network/pull/666
Hole-punching: https://github.com/maidsafe/safe_network/pull/667

**Pull request closed without merging**